### PR TITLE
[DEV] switch nose with pytest

### PR DESCRIPTION
**Pull request by szha**
## Description ##
switch nose with pytest

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] switch nose with pytest

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by leezu** [comment]
How about the random seed handling and reproducibility? I think you'll need to adapt https://github.com/dmlc/gluon-nlp/blob/v0.9.x/conftest.py#L41

**Comment by szha** [comment]
@leezu thanks. Added.

File: .github/workflows/os_x_staticbuild.yml [hunk 1]
```diff
@@ -10,16 +10,16 @@ jobs:
         uses: actions/checkout@v2
       - name: Install Dependencies
         run: |
-          brew install nasm automake ninja libtool cmake pkgconfig protobuf
+          brew install nasm automake ninja libtool cmake pkgconfig protobuf hdf5 zlib
+          python3 -m pip install --user -r ci/docker/install/requirements
```
Comment (leezu) [inline .github/workflows/os_x_staticbuild.yml:14]: I suggest not to use `ci/docker/install/requirements` here but list the dependencies explicitly. Most of the things in `ci/docker/install/requirements` are not needed for this test and having a single requirements file used for a that is used from an opaque number of locations for an variety of purposes makes it very hard to maintain that file.

File: .github/workflows/os_x_staticbuild.yml [hunk 2]
```diff
@@ -10,16 +10,16 @@ jobs:
         uses: actions/checkout@v2
       - name: Install Dependencies
         run: |
-          brew install nasm automake ninja libtool cmake pkgconfig protobuf
+          brew install nasm automake ninja libtool cmake pkgconfig protobuf hdf5 zlib
```
Comment (leezu) [inline .github/workflows/os_x_staticbuild.yml:13]: Why are the dependencies on hdf5 zlib introduced when switching to pytest?

Comment (szha) [reply .github/workflows/os_x_staticbuild.yml:13]: this was for installing h5py and pillow, which triggers compile in source distribution.

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-cpu, windows-cpu]

File: conftest.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+    The segfault-debug procedure on a module called test_module.py is:
+
+    1. run "pytest --verbose test_module.py".  A seg-faulting output might be:
+
+       [INFO] np, mx and python random seeds = 4018804151
+       test_module.test1 ... ok
+       test_module.test2 ... Illegal instruction (core dumped)
+
+    2. Copy the module-starting seed into the next command, then run:
```
Comment (marcoabreu) [inline conftest.py:52]: Would it be possible to simplify this process? It seems like multiple iterations would be required to get the correct seed.

Instead, would it be possible to print the used seeds after a test failed?

File: conftest.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+          'use MXNET_MODULE_SEED={} to reproduce.'.format(seed))
+
+    random.seed(seed)
+    try:
+        import numpy as np
+        import mxnet as mx
+        np.random.seed(seed)
+        mx.random.seed(seed)
+    except:
+        print('Unable to import numpy/mxnet. Skipping conftest.')
```
Comment (marcoabreu) [inline conftest.py:103]: Inconsistent use of print/logging

File: conftest.py (hunk 3 above)
Comment (marcoabreu) [reply conftest.py:52]: Ah, it seems like the test seed value is actually printed as I requested here: https://github.com/apache/incubator-mxnet/pull/18025/files#diff-c9b030b0828c1291b56fc10aba6baab6R204

Is that correct?

File: conftest.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    elif env_seed_str is not None:
+        seed = int(env_seed_str)
+    else:
+        seed = random.randint(0, 2**31-1)
+
+    random.seed(seed)
+    try:
+        import numpy as np
+        import mxnet as mx
+        post_test_state = np.random.get_state()
```
Comment (marcoabreu) [inline conftest.py:173]: what about the mx.random state? Are we unable to save it?

**Comment by marcoabreu** [comment]
Just checked a sample pytest run in CI. Love it! Great initiative, Sheng!

File: conftest.py (hunk 3 above)
Comment (szha) [reply conftest.py:52]: Yes, it's already logged.

File: conftest.py (hunk 5 above)
Comment (szha) [reply conftest.py:173]: yes, we are unable to save it yet as we don't expose the states in the frontend yet.

File: conftest.py (hunk 4 above)
Comment (szha) [reply conftest.py:103]: switched all prints to logging.info/warning

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by szha** [comment]
I can see the retry job triggered by the CI bot and the job running, but the state doesn't seem to be reflected in the check status on github.

**Comment by leezu** [comment]
The state get's updated after the run finishes. 

cc @ChaiBapchya who may fix the bot to display "in progress" state instead of failed state in this situation.

**Comment by marcoabreu** [comment]
The state should not be published by the bot but by CI. It should report in progress if it has been retriggered. Only exception might be if a new run has been requested before the existing one finished. In that case, the successful one might have overridden the in progress state because it happened later. 

It should be investigated, but not as part of the bot.

**Comment by leezu** [comment]
> In that case, the successful one might have overridden the in progress state because it happened later.

@marcoabreu That's the point. The bot is responsible for the order of events thus the bot needs to be investigated. The issue does not appear when manually stopping and then retriggering a job on CI. The bot needs to allow time for Github to report the failed state of the old job before starting the new job.

**Comment by szha** [comment]
and github was down briefly 🤦‍♂️

**Comment by leezu** [comment]
> An option would of course be to cancel the active run, but that's a separate question I guess.

This is exactly what is done already. The problem really appears to be the timing of cancelling and starting the new job

**Comment by marcoabreu** [comment]
Would you mind not force pushing so I can review the changes as we go?

**Comment by szha** [comment]
@marcoabreu sure, I will if feasible. The last one was for resolving a conflict. Unfortunately, I'm about to make another force push due to new conflicts.

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by marcoabreu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by leezu** [comment]
@szha there is a bug with the unix-gpu dist kvstore tests

Runtime on master is 6 minutes http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/master/1887/pipeline/426 but more than 1 hour in this PR http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-18025/56/pipeline

**Comment by leezu** [comment]
Ok, the difference is due to hanging issue..

[2020-04-22T17:06:49.477Z] AssertionError: (None, array([[0.6, 0.6, 0.6],

[2020-04-22T17:06:49.477Z]        [0.6, 0.6, 0.6]], dtype=float32), 41.0)

[2020-04-22T17:06:49.477Z] terminate called without an active exception

[2020-04-22T17:06:49.477Z] terminate called without an active exception

[2020-04-22T17:06:49.477Z] terminate called without an active exception

[2020-04-22T17:06:49.477Z] terminate called without an active exception

Thanks @marcoabreu to retrigger

**Comment by marcoabreu** [comment]
The stability of our test suite is really a pain in the butt :(

You're welcome

**Comment by leezu** [comment]
This does not seem to be a stability issue, but a bug. Same failure again:

```
[2020-04-22T19:21:39.589Z] Traceback (most recent call last):

[2020-04-22T19:21:39.589Z]   File "dist_device_sync_kvstore.py", line 130, in <module>

[2020-04-22T19:21:39.589Z]     test_sync_push_pull()

[2020-04-22T19:21:39.589Z]   File "dist_device_sync_kvstore.py", line 88, in test_sync_push_pull

[2020-04-22T19:21:39.589Z]     check_default_keys(kv, my_rank, nworker, nrepeat=3)

[2020-04-22T19:21:39.589Z]   File "dist_device_sync_kvstore.py", line 72, in check_default_keys

[2020-04-22T19:21:39.589Z]     check_diff_to_scalar(val, num)

[2020-04-22T19:21:39.589Z]   File "dist_device_sync_kvstore.py", line 29, in check_diff_to_scalar

[2020-04-22T19:21:39.589Z]     assert(np.sum(np.abs((A - x).asnumpy())) == 0), (rank, A.asnumpy(), x)

[2020-04-22T19:21:39.589Z] AssertionError: (None, array([[0.6, 0.6, 0.6],

[2020-04-22T19:21:39.589Z]        [0.6, 0.6, 0.6]], dtype=float32), 41.0)
```

@szha

**Comment by leezu** [comment]
Still

```
[2020-04-22T21:11:19.922Z] + cd /work/mxnet/tests/nightly

[2020-04-22T21:11:19.922Z] + ./test_distributed_training-gpu.sh

[2020-04-22T21:11:41.844Z] worker 0 is initialized

[2020-04-22T21:11:41.844Z] worker 0 passed test_gluon_trainer_type

[2020-04-22T21:11:41.844Z] worker 2 is initialized

[2020-04-22T21:11:41.844Z] worker 2 passed test_gluon_trainer_type

[2020-04-22T21:11:41.844Z] worker 1 is initialized

[2020-04-22T21:11:41.844Z] worker 1 passed test_gluon_trainer_type

[2020-04-22T21:11:41.844Z] worker 3 is initialized

[2020-04-22T21:11:41.844Z] worker 3 passed test_gluon_trainer_type

[2020-04-22T21:11:41.844Z] terminate called without an active exception

[2020-04-22T21:11:41.844Z] terminate called without an active exception

[2020-04-22T21:11:41.844Z] terminate called without an active exception

[2020-04-22T21:11:41.844Z] terminate called without an active exception

```

File: ci/jenkins/Jenkinsfile_centos_cpu [hunk 6]
```diff
@@ -21,7 +21,7 @@
 // See documents at https://jenkins.io/doc/book/pipeline/jenkinsfile/
 
 // timeout in minutes
-max_time = 180
+max_time = 240
```
Comment (leezu) [inline ci/jenkins/Jenkinsfile_centos_cpu:24]: why increase the timeout? http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-18025/59/pipeline/ 

Now the jobs hang even longer before termination

Comment (szha) [reply ci/jenkins/Jenkinsfile_centos_cpu:24]: I noticed that windows tests were running into timeout. also, we should have other ways for handling hanging issue than relying on step level timeout.

**Comment by marcoabreu** [comment]
Woohoo!

File: ci/docker/runtime_functions.sh [hunk 7]
```diff
@@ -994,26 +991,24 @@ cd_unittest_ubuntu() {
...
+    pytest --durations=50 --verbose tests/python/quantization
 
     # https://github.com/apache/incubator-mxnet/issues/11801
     # if [[ ${mxnet_variant} = "cpu" ]] || [[ ${mxnet_variant} = "mkl" ]]; then
         # integrationtest_ubuntu_cpu_dist_kvstore
     # fi
 
     if [[ ${mxnet_variant} = cu* ]]; then
-        $nose_cmd $NOSE_TIMER_ARGUMENTS --verbose tests/python/gpu
+        pytest --durations=50 --verbose tests/python/gpu
```
Comment (ChaiBapchya) [inline ci/docker/runtime_functions.sh:1003]: why are we showing top 50 slowest duration only?

Comment (marcoabreu) [reply ci/docker/runtime_functions.sh:1003]: Why not? The others are not really of interest

**Pull request merged**